File: functional.scala
object Functional {

  def fac(n : Int) : Int = {
    @annotation.tailrec
    def go(n : Int, acc : Int) : Int = {
      if(n <= 0) acc
      else go(n-1,n*acc)
    }
    go(n,1)
  }

  def isSorted[A](as : Array[A], f : (A,A) => Boolean) : Boolean = {
    def loop(n : Int) : Boolean = {
      if(as.length < 2) false
      else if(n >= as.length - 2 ) f(as(n),as(n+1))
      else if (!f(as(n),as(n+1))) false
      else loop(n+1)
    }
    loop(0)
  }

  def findFirst[A](ss : Array[A], p: A => Boolean) : Int = {
    @annotation.tailrec
    def loop(n : Int) : Int = {
      if(n >= ss.length) -1
      else if(p(ss(n))) n
      else loop(n+1)
    }
    loop(0)
  }

  def partial[A,B,C](a : A, f:(A,B) => C): B => C =
    (b : B) => f(a,b)

  def curry[A,B,C](f: (A,B) => C): A => (B => C) = {
    (a : A) => partial1(a,f)
  }
  
  def uncurry[A,B,C](f: A => B => C) : (A, B) => C = {
    (a : A, b : B) => f(a)(b)
  }

  def compose[A,B,C](f : B => C, g : A => B ) : A => C = {
    (a : A) => f(g(a))
  }

}

File: functional.cpp
#include <functional>
#include <iostream>

template <typename A, typename B, typename C>
auto partial(A a, const std::function<C(A,B)>& f) 
{
    return std::function<C(B)>([=](B b) { return f(a, b); });
}

template <typename A, typename B, typename C>
auto curry(const std::function<C(A,B)>& f)
{
    return std::function<std::function<C(B)>(A)>([=](A a) { return partial(a, f); });
}

template <typename A, typename B, typename C>
auto uncurry(const std::function<std::function<C(B)>(A)>& f)
{
    return std::function<C(A, B)>([=](A a, B b) { return f(a)(b); });
}

template <typename A, typename B, typename C>
auto compose(const std::function<C(B)>& f, const std::function<B(A)>& g)
{
    return std::function<C(A)>([=](A a) { return f(g(a)); });
}

int main()
{
    std::function<int(int, int)> f([](int a, int b) { return a + b; });
    std::function<int(int, int)> g([](int a, int b) { return a + b; });

    std::function<int(int)> f1([](int b) { return b + 10; });
    std::function<int(int)> g1([](int a) { return a + 20; });

    std::cout << partial(10, f)(100) << std::endl;
    std::cout << uncurry(curry(g))(10,20) <<std::endl;
    std::cout << compose(f1,g1)(10) << std::endl;
}


